File: VersionJava/Cocinero.java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

public class Cocinero extends Thread {
    private final List<Pedido> pedidos;
    private final Object lock;
    private final File logFile;

    public Cocinero(String nombre, List<Pedido> pedidos, Object lock, File logFile) {
        super(nombre);
        this.pedidos = pedidos;
        this.lock = lock;
        this.logFile = logFile;
    }

    @Override
    public void run() {
        while (true) {
            Pedido pedido = null;

            synchronized (lock) {
                if (pedidos.isEmpty()) {
                    return; // No hay más pedidos
                }
                pedido = pedidos.remove(0);
            }

            // Simula preparación
            double tiempoPreparacion = 1 + Math.random() * 2; // 1 a 3 segundos
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS");
            String timestamp = LocalDateTime.now().format(formatter);
            System.out.printf("[%s] %s está preparando %s... (%.1fs)%n", timestamp, getName(), pedido.getNombre(), tiempoPreparacion);

            try {
                Thread.sleep((long)(tiempoPreparacion * 1000));
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

            // Registrar en log
            synchronized (lock) {
                try (BufferedWriter bw = new BufferedWriter(new FileWriter(logFile, true))) {
                    String linea = String.format("[%s] %s ha preparado %s", 
                            timestamp, getName(), pedido.getNombre());
                    bw.write(linea);
                    bw.newLine();
                    bw.flush();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: version_python/cocinero.py
import os
import threading
import time
import random
from datetime import datetime

class Cocinero(threading.Thread):
    """
    Cada cocinero es un hilo que procesa pedidos de una lista compartida.
    Usa un Lock para acceder de forma segura a la lista y al archivo de log.
    """

    def __init__(self, nombre: str, pedidos: list, lock: threading.Lock, log_file: str):
        super().__init__(name=nombre)
        self.pedidos = pedidos
        self.lock = lock
        self.log_file = log_file

    def run(self):
        while True:
            self.lock.acquire()
            try:
                if not self.pedidos:
                    # No hay más pedidos que procesar
                    return
                pedido = self.pedidos.pop(0)
            finally:
                self.lock.release()

            # Simula el tiempo de preparación del pedido
            tiempo_preparacion = random.uniform(1, 3)
            
            timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
            print(f"[{timestamp}] {self.name} está preparando {pedido.plato}... ({tiempo_preparacion:.1f}s)")
            time.sleep(tiempo_preparacion)

            # Registrar en el log (zona crítica protegida)
            with self.lock:
                timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
                linea = f"[{timestamp}] {self.name} ha preparado {pedido.plato} (Pedido {pedido.id})"
                with open(self.log_file, "a", encoding="utf-8") as f:
                    f.write(linea + "\n")
